Objectives functionality overhaul

MadmanMartian opened the pull request:
:cl: 
tweak: More objectives are generated than just the one. The amount of objectives generated at a time can be toggled in the config.
tweak: The bomb or papers delivery objective now requires you to move the bomb or papers into the cargo selling spot, and press delivery on the cargo console.
tweak: The cargo console will only show things to sell and or deliver should there be any actually on the spot for selling and or delivering
/:cl:

[why]: # People were asking for it.

MadmanMartian committed:
More objectives are generated than just the one at a time.

MadmanMartian commented:
Going to remake according to Altangy's suggestion

 - Should only generate the multiple objectives at roundstart, then when one is completed it adds another.

MadmanMartian committed:
Whoops I accidentally rewrote how station objectives were handled.

MadmanMartian commented:
Dang. Compiles on 512.1437. I'll rework the check.

MadmanMartian committed:
Compiles under 511.1380

MadmanMartian commented:
welp. Never saved changes to the cargo UI.

MadmanMartian committed:
Hal-lightbulb-fixing.gif

MadmanMartian commented:
I hereby despise TGUI

Qustinnus commented:
welcome to the club this is why im reluctant to update combat and events

MadmanMartian committed:
After a whole week, it works

TMTIME commented on code/modules/cargo/console.dm:
```diff
@@ -152,6 +153,11 @@ proc/get_cost_multiplier(var/datum/planet/PL)
 						"cost" = price,// / cost_mult,
 						"id" = "\ref[O]"
 					))
+				data["deliver"] += list(list(
```
Volatile Bombs won't reach here because of the continue on line 143. Might want to also make deliver only show for mission items.

TMTIME commented on code/__HELPERS/objectives.dm:
```diff
@@ -0,0 +1,37 @@
+		if(!go_home && (O.failed || O.completed))
+			generate_ship_objective(TRUE)
+	if(go_home && !locate(/datum/objective/ftl/gohome) in get_ship_objectives())
+		add_ship_objective(new /datum/objective/ftl/gohome)
```
Need that true

TMTIME reviewed (changes requested):
Other than this, seems alright

MadmanMartian committed:
Requested changes

 - Moves the target objective targets to a global list for access by the target objective and the cargo system
 - The gohome objective now informs the crew it is time to go home
 - The sell and deliver options now only appear should there be anything to deliver and/or sell to the station.

MadmanMartian commented:
Found that, for some weird reason, buying a limited stock item doesn't drop the amount that the station has.

TMTIME commented on code/__HELPERS/objectives.dm:
```diff
@@ -0,0 +1,37 @@
+		go_home = TRUE
+	for(var/datum/objective/O in get_ship_objectives())
+		O.check_completion()
+		if(!go_home && (O.failed || O.completed))
```
Selling any item with atleast one completed objective will generate a new objective. Maybe check to see if an objective isn't finished before checking completion -> new objective.

MadmanMartian replied:
okay, so you want it so that once you've completed all objectives, you then start to get more objectives?

TMTIME replied:
like you;ve done aa a a a

TMTIME commented on code/game/gamemodes/objective.dm:
```diff
@@ -907,52 +908,57 @@ GLOBAL_LIST_EMPTY(possible_items_special)
+				target_planet = SSstarmap.pick_station("syndicate") //I don't get why we haul a bomb from NT to NT, so lets take it to the Syndicate
+				if(!istype(target_planet.parent_system,/datum/star_system/capital/syndicate)) //Dolos check
+					searching_planets = FALSE
+		if("NT2NT") //Pizza delivery
```
do it i dare ya

TMTIME commented on code/modules/cargo/console.dm:
```diff
@@ -164,6 +171,15 @@ proc/get_cost_multiplier(var/datum/planet/PL)
+				return
+			if(O.invisibility >= INVISIBILITY_ABSTRACT || O.anchored)
+				return
+			if(is_type_in_list(O, GLOB.objective_delivery_types))  //TODO: Add these to a global variable
```
Not a todo anymore

TMTIME reviewed (dismissed):
the pizza mission is optional tbh

MadmanMartian committed:
removes todo

MadmanMartian committed:
penispenispenis

MadmanMartian committed:
Apparently O.completed is not set by check_completion in some cases

MadmanMartian committed:
Forgot boarding! Thanks TMTIME

TMTIME commented:
Hold System won't work either. 

Since its shitcode feel free to just disable it for now and I'll fix it later.

If you want to fix it, you'll need to convert all its world.time stuff to addtimers.
Also checking_objectives when arriving at a system will help the gohome objective finish + this one start

TMTIME committed:
Cleans up hold objective and checks objectives after jump

Qustinnus commented:
aaa i know nothing about objectives

TMTIME commented:
with a boarding objective underway, killing a ship seems to have procced make a new objective

MadmanMartian commented:
The objective was to go home

MadmanMartian commented on code/__HELPERS/objectives.dm:
```diff
@@ -0,0 +1,39 @@
+/proc/check_ship_objectives()
+	var/go_home = FALSE
+	if(world.time > 54000 || !get_ship_objective_types_len())
```
This would be the cause. Because world.time is greater than 54000

TMTIME commented:
If a go home objective is active and the crew completes an objective, they receive no announcement if they complete another objective

MadmanMartian commented:
That doesn't sound right

MadmanMartian committed:
Adds yet more config options, and more player feedback.

MadmanMartian commented:
Latest pull:
 - Adds another config option: How many objectives MUST be completed before you can go home.
 - Adds some user feedback for when an objective is added (IE, if an objective was completed/failed, and if it's time to go home)

crossedfall commented:
All tests have been fine. Aside from some minor feedback which can be addressed in a separate PR, I think this would be good to merge. @Qustinnus
